# FastText Embedding from Eng_to_Ger and Ger_to_Eng

All datasets and embeddings are present in this drive and hence I have mount it in this colab file to use them as per my need

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Downloading word2vec skipgram embeddings for English and German and unzipping in drive folder

In [ ]:
!wget http://vectors.nlpl.eu/repository/20/40.zip


--2021-04-07 12:12:43--  http://vectors.nlpl.eu/repository/20/40.zip
Resolving vectors.nlpl.eu (vectors.nlpl.eu)... 129.240.189.181
Connecting to vectors.nlpl.eu (vectors.nlpl.eu)|129.240.189.181|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3033545498 (2.8G) [application/zip]
Saving to: ‘40.zip’

40.zip              100%[===================>]   2.83G  10.0MB/s    in 4m 56s  

2021-04-07 12:17:41 (9.79 MB/s) - ‘40.zip’ saved [3033545498/3033545498]



In [ ]:
!unzip /content/40.zip

Archive:  /content/40.zip
  inflating: LIST                    
  inflating: meta.json               
  inflating: model.bin               
  inflating: model.txt               
  inflating: README                  


In [ ]:
!wget http://vectors.nlpl.eu/repository/20/45.zip

--2021-04-07 12:38:51--  http://vectors.nlpl.eu/repository/20/45.zip
Resolving vectors.nlpl.eu (vectors.nlpl.eu)... 129.240.189.181
Connecting to vectors.nlpl.eu (vectors.nlpl.eu)|129.240.189.181|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3764660264 (3.5G) [application/zip]
Saving to: ‘45.zip’

45.zip              100%[===================>]   3.51G  10.2MB/s    in 6m 2s   

2021-04-07 12:44:55 (9.91 MB/s) - ‘45.zip’ saved [3764660264/3764660264]



In [ ]:
!unzip /content/45.zip



Archive:  /content/45.zip
  inflating: LIST                    
  inflating: meta.json               
  inflating: model.bin               
  inflating: model.txt               
  inflating: README                  


In [2]:
cd /content/drive/MyDrive/Lab_Task_7

/content/drive/MyDrive/Lab_Task_7


Cloning the VecMap library to generate bilingual dictionary

In [3]:
!git clone https://github.com/artetxem/vecmap.git

fatal: destination path 'vecmap' already exists and is not an empty directory.


In [4]:
cd /content/drive/MyDrive/Lab_Task_7/vecmap

/content/drive/MyDrive/Lab_Task_7/vecmap


In order to build cross-lingual word embeddings, we will first train monolingual word embeddings for each language using fasttext and then map them to a common space with VecMap software. Having done that, we can evaluate the resulting cross-lingual embeddings using repo's included tools.

In [5]:
!python3 map_embeddings.py --supervised ../Train_ger_eng.txt ../german.vec ../eng.vec SRC_MAPPED.EMB TRG_MAPPED.EMB

tcmalloc: large alloc 1656438784 bytes == 0x555dbe432000 @  0x7f9c8fd531e7 0x7f9c8d79346e 0x7f9c8d7e3c7b 0x7f9c8d7e435f 0x7f9c8d886103 0x555d1e2090e4 0x555d1e208de0 0x555d1e27d6f5 0x555d1e277e0d 0x555d1e20a77a 0x555d1e278a45 0x555d1e277b0e 0x555d1e277813 0x555d1e341592 0x555d1e34190d 0x555d1e3417b6 0x555d1e319103 0x555d1e318dac 0x7f9c8eb3dbf7 0x555d1e318c8a
tcmalloc: large alloc 2627141632 bytes == 0x555e20fe6000 @  0x7f9c8fd531e7 0x7f9c8d79346e 0x7f9c8d7e3c7b 0x7f9c8d7e435f 0x7f9c8d886103 0x555d1e2090e4 0x555d1e208de0 0x555d1e27d6f5 0x555d1e277e0d 0x555d1e20a77a 0x555d1e278a45 0x555d1e277b0e 0x555d1e277813 0x555d1e341592 0x555d1e34190d 0x555d1e3417b6 0x555d1e319103 0x555d1e318dac 0x7f9c8eb3dbf7 0x555d1e318c8a


For German to English, we will now evaluate our mapped embeddings in bilingual lexicon extraction (aka dictionary induction or word translation) as follows:

In [6]:
!python3 eval_translation.py SRC_MAPPED.EMB TRG_MAPPED.EMB -d ../german_to_eng_dict_multiple.txt --retrieval csls --cuda

Coverage:100.00%  Accuracy: 69.94%


Accuracy comes out to be **69.94%**

We will now train monolingual word embeddings for each language using fasttext and then map them to a common space with VecMap software : for English to German

In [7]:
!python3 map_embeddings.py --supervised ../Train_eng_ger.txt ../eng.vec ../german.vec ENG_MAPPED.EMB GER_MAPPED.EMB 

tcmalloc: large alloc 2627141632 bytes == 0x55c82b7e0000 @  0x7f64ec53b1e7 0x7f64e9f7b46e 0x7f64e9fcbc7b 0x7f64e9fcc35f 0x7f64ea06e103 0x55c78be5f0e4 0x55c78be5ede0 0x55c78bed36f5 0x55c78becde0d 0x55c78be6077a 0x55c78becea45 0x55c78becdb0e 0x55c78becd813 0x55c78bf97592 0x55c78bf9790d 0x55c78bf977b6 0x55c78bf6f103 0x55c78bf6edac 0x7f64eb325bf7 0x55c78bf6ec8a
tcmalloc: large alloc 1656438784 bytes == 0x55c8c8150000 @  0x7f64ec53b1e7 0x7f64e9f7b46e 0x7f64e9fcbc7b 0x7f64e9fcc35f 0x7f64ea06e103 0x55c78be5f0e4 0x55c78be5ede0 0x55c78bed36f5 0x55c78becde0d 0x55c78be6077a 0x55c78becea45 0x55c78becdb0e 0x55c78becd813 0x55c78bf97592 0x55c78bf9790d 0x55c78bf977b6 0x55c78bf6f103 0x55c78bf6edac 0x7f64eb325bf7 0x55c78bf6ec8a


For English to German, we will now evaluate our mapped embeddings in bilingual lexicon extraction (aka dictionary induction or word translation) as follows:

In [8]:
!python3 eval_translation.py ENG_MAPPED.EMB GER_MAPPED.EMB -d ../high_word_pairs.txt --retrieval csls --cuda

Coverage: 94.10%  Accuracy: 75.56%


Accuracy comes out to be **75.56%**